File: 백준/Gold/16929. Two Dots/Two Dots.java
import java.util.*;
import java.io.*;

public class Main {
    static int n, m;
    static char[][] board;
    static boolean[][] visited;
    static int[] dr = {-1, 1, 0, 0};
    static int[] dc = {0, 0, -1, 1};

    public static void dfs(int s_r, int s_c, int r, int c, int cnt) {
        for (int k = 0; k < 4; k++) {
            int nr = r + dr[k];
            int nc = c + dc[k];

            if (nr < 0 || nr >= n || nc < 0 || nc >= m) {
                continue;
            } else if (!(visited[nr][nc]) && board[nr][nc] == board[r][c]) {
                visited[nr][nc] = true;
                dfs(s_r, s_c, nr, nc, cnt+1);
                visited[nr][nc] = false;
            } else if (cnt >= 4 && nr == s_r && nc == s_c) {
                System.out.println("Yes");
                System.exit(0);
                return;
            }
        }
    }

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        board = new char[n][m];
        visited = new boolean[n][m];

        // 배열 입력받기
        for (int i = 0; i < n; i++) {
            String str = br.readLine();
            for (int j = 0; j < m; j++) {
                board[i][j] = str.charAt(j);
            }
        }

        // dfs 실행
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                visited[i][j] = true;
                dfs(i, j, i, j, 1);
                visited[i][j] = false;
            }
        }

        System.out.println("No");
    }
}

File: 백준/Gold/16929. Two Dots/Two Dots.py
# dfs 돌면서 idx가 4이상일때 다시 원점으로 돌아오면 사이클이 있는 것
import sys
input = sys.stdin.readline
N, M = map(int, input().split())
board = [list(input()) for _ in range(N)]

find = False
visited = [[False] * M for _ in range(N)]

dr = [-1, 1, 0, 0]
dc = [0, 0, -1, 1]

def dfs(s_r, s_c, i, j, idx):
    for k in range(4):
        nr = i + dr[k]
        nc = j + dc[k]
        if 0 <= nr < N and 0 <= nc < M:
            if not visited[nr][nc] and board[nr][nc] == board[i][j]:
                visited[nr][nc] = True
                dfs(s_r, s_c, nr, nc, idx+1)
                visited[nr][nc] = False
            elif s_r == nr and s_c == nc and idx >= 4:
                print("Yes")
                exit()

for i in range(N):
    for j in range(M):
        visited[i][j] = True
        dfs(i, j, i, j, 1)
        visited[i][j] = False


print("No")
